Refactoring the generate() function

patrickvonplaten opened the pull request:
## Generation refactor

This is a possible design that IMO would make generate much more readable and flexible for future code changes. 
The code shown in this PR is more or less pseudo code, but I'm sure that this design is fully backwards compatible (for the `generate()` function, I don't plan on keep `beam_search_generation` and `_no_beam_search_generation` -> I don't think anybody directly accessed these functions).

It's probably better to look at the code directly, here: https://github.com/huggingface/transformers/blob/3c8a12a3bf44aa7845675852e28c47d0a9cb808e/src/transformers/generation_utils.py then at the diff.

The following major changes are made:

### Split the generate method into four generate functions:

a) `greedy_search` (corresponds to num_beams = 1, do_sample=False)
b) `sample` (corresponds to num_beams = 1, do_sample=True)
c) `beam_search` (corresponds to num_beams > 1, do_sample = False)
d) `beam_sample` (corresponds to num_beams > 1, do_sample = True)

It is split in a way that the functions can be used on their own and don't necessarily have to be accessed by the `generate()` method. This allows for much more flexibility for the user - he can decide what kind of distribution warper he wants to use and what kind of "Beam scorer" he wants to use (more on "distribution warper" and "beam scorer" later.
     Also mostly a model only uses one of these methods, `EncoderDecoder` models usually use `beam_search()`, `...ForCausalLM` models usually use `sample()`. Because `generate` is now split to the corresponding functions relevant for a model, it makes the code much more readable for the users because they will mostly only look into  one of the four functions.
    Splitting `generate()` into four functions removes **a lot** of `if-else` statements

### Create `LogitsProcessor` and `LogitsWarper` objects. 

Instead of adding each of these "logit warpers / processors", such as `bad_token_words` with `if-else` statements, a list of these objects is created in the beginning and then called in the code. This is largely copied from this very nice PR: https://github.com/huggingface/transformers/pull/5420 . The arguments are 1) easier to test, 2) adds more flexbility because users can easily add their own "logit warpers / processors". 
Note that we need both a `pre_processor` and a `dist_warper` list to make `beam_sample` work correctly. This comment explains why in more detail: https://github.com/huggingface/transformers/pull/5420#discussion_r449779867

### Create `beam_scorer` class.

The `generate_beam_search` function has become extremely hard to read because most of the beam search logic is directly written into the function. I would propose to move this code into a `beam_scorer` class that would also replace the `BeamHypotheses` class. The class would essentially expose the following functions: 

```python
beam_scorer.update(next_scores, next_tokens)

next_beam_scores = beam_scorer.get_next_scores()
next_beam_tokens = beam_scorer.get_next_tokens()
next_beam_idx = beam_scorer.get_next_beam_idx()

beam_scorer.is_done()
```

and IMO all the beam search relevant code can be handled in the `beam_scorer.update(...)` function.
Besides better readability, such a class could also be replaced by another "Beam seach scorer" logic, which makes the beam search code much easier to extend IMO.

The con of this PR is obviously that some code would be copy pasted along the four functions (but it should not be too much).

Let me know if something is not clear in the design. Would be super happy to hear your feedback @LysandreJik @thomwolf @sshleifer @yjernite @JetRunner @sgugger 

## 1st Review:

The current state shows how the refactored code would look like. All important tests are passing now (the slow GPT2, Bart & T5 tests), but the PR is not at all finished yet. It would be nice if @sshleifer @yjernite @thomwolf (and @LysandreJik @sgugger) could take a first look at the complete "new" architecture and give some feedback. If you guys are ok with the new design, I will add a bunch of tests and clean up the code more to make sure we have 100% backward compatibily. 

## TODO:
- [x] Make all slow tests pass
- [x] Rename according to discussion
- [x] Add tests for all processors
- [x] Add more and better generation tests
- [x] Do speed comparison
- [x] Add docstring
- [x] Final thoughts about design

## Final review:

I'm happy with the changes I've done now. Better tests (more aggressive and should also be faster since only 3 tokens are generated instead of 10 now) have been added and docstrings have hopefully become more understandable. 
A complete explanation of new "generation" philosophy is described in the forum: https://discuss.huggingface.co/t/big-generate-refactor/1857

patrickvonplaten committed:
first draft

patrickvonplaten committed:
show design proposition for new generate method

patrickvonplaten committed:
up

patrickvonplaten committed:
make better readable

sshleifer commented:
I think this is a massive improvement and really hope it gets merged. Everything is much better encapsulated and it's super easy to find what you want.

I have some naming ideas, but the code seems to not be finished so I will wait.

sgugger commented:
I am very unfamiliar with the generate code because the code has scared me way, this one is way more inviting to read and I can actually understand it so I think that's a very nice achievement :-)

I don't know if what's hidden in `# add necessary encoder decoder code` and `# add all post processing functions` is different for the four versions of decode, but might be nice to refactor it in some helper methods if possible.

yjernite commented:
Seconding Sam and Sylvain. Really excited for the improved legibility but we should also make sure that code isn't copy-pasted four times or we'll start having inconsistencies soon :) 

We should probably allow `dist_warper` in the beam search too: the current ones wouldn't make a difference but some use cases will (e.g. noisy channel which uses a backward distribution to reweigh the next token scores)

patrickvonplaten commented:
Related: https://github.com/huggingface/transformers/issues/7626

patrickvonplaten committed:
make first version

patrickvonplaten committed:
fix merge conflicts

patrickvonplaten committed:
gpt2 tests pass

patrickvonplaten committed:
Merge branch 'master' of https://github.com/huggingface/transformers into generation_refactor

patrickvonplaten committed:
make beam search for gpt2 work

patrickvonplaten committed:
add first encoder-decoder code

LSinev commented:
Can you please add this functionality too https://github.com/huggingface/transformers/issues/5164?

patrickvonplaten committed:
delete typo

patrickvonplaten committed:
make t5 work

patrickvonplaten committed:
save indermediate

patrickvonplaten committed:
:wqaMerge remote-tracking branch 'main/master' into generation_refactor

patrickvonplaten committed:
make bart work with beam search

patrickvonplaten committed:
finish beam search bart / t5

patrickvonplaten committed:
add default kwargs

sshleifer commented on src/transformers/generation_utils.py:
```diff
@@ -128,843 +189,449 @@ def generate(
+            length_penalty = length_penalty if length_penalty is not None else self.config.length_penalty
+            early_stopping = early_stopping if early_stopping is not None else self.config.early_stopping
+
+            beam_scorer = BeamSearchBase(
```
maybe class should be `BeamSearchScorer`

sshleifer commented on src/transformers/generation_utils.py:
```diff
@@ -128,843 +189,449 @@ def generate(
+        return sequence_lengths, unfinished_sequences
+
+    @staticmethod
+    def update_model_kwargs(outputs, model_kwargs, is_encoder_decoder=False):
```
this is named to generally I think, I missed the caller but is this different than `prepare_inputs_for_generation`?

Should models that use `mems` overwrite?

patrickvonplaten replied:
`mems` are also overwritten here, as well as the attention mask. In the future things like `position_ids`, `token_type_ids` might join as well. I could append a `_for_generation` to make it clearer.

sshleifer commented on src/transformers/generation_utils.py:
```diff
@@ -128,843 +189,449 @@ def generate(
-    @staticmethod
-    def _reorder_cache(past: Tuple, beam_idx: Tensor) -> Tuple[Tensor]:
-        return tuple(layer_past.index_select(1, beam_idx) for layer_past in past)
+            # adjust tokens for Bart, *e.g.*
```
```suggestion
            # adjust token scores (a no-op by default)
```

sshleifer reviewed (commented):
LGTM, left some nits.

and wrote you a new integration test alias:

```bash
run_generation_integration_tests () {
	RUN_SLOW=1 pytest tests/test_modeling_pegasus.py
	RUN_SLOW=1 pytest tests/test_modeling_bart.py
	RUN_SLOW=1 pytest tests/test_modeling_t5.py
	RUN_SLOW=1 pytest tests/test_modeling_marian.py
	RUN_SLOW=1 pytest tests/test_modeling_mbart.py
	RUN_SLOW=1 pytest tests/test_modeling_blenderbot.py
	RUN_SLOW=1 pytest tests/test_modeling_fsmt.py
}
```

yjernite commented:
This is really cool! I love the additional clarity and think you're building a really strong base for the generation code. Here's what I think we should still chat about before finalizing the PR:
1. We still have inconsistencies in how we handle the end of generation. I really think that `max_length` should be handled the same way as `min_length` with a sampler in `pre_processor ` which forces a Dirac on `eos_token_id` when `cur_len==max_length`
2. Similarly, we should get rid of `adjust_logits_during_generation` and have e.g. a Bart-specific sampler in `pre_processor `
3. We're missing out on supporting some really interesting research by not giving a good option to return and backprop through the generation scores. Also, I think we should extend self-documenting outputs to the `generate` function (cc @sgugger ). My proposal would be to add a `return_dict` argument, with an `output_generation_scores` (and possibly `with_grad`) option. (And we just return the generated ids if `return_dict=False` to stay backwards compatible)
4. Generation with `decoder_prefix_ids` for encoder-decoder models :D ! That can be a future PR though, and can also be handled with `pre_processor` to force the output (a bit wasteful but will do in a pinch)

I also agree with @sgugger that we should come up with a better name than `Sampler`, will think about it

sshleifer commented:
@yjernite has great feature requests, but this PR is already huge and I don't see why they need to be handled here.

yjernite commented:
> @yjernite has great feature requests, but this PR is already huge and I don't see why they need to be handled here.

Good point, we can definitely look at 3. and 4. later, and I know from experience that 2. is probably the wrong kind of rabbit hole to get into right now.

I am a little concerned about 1. though. The samplers will definitely need to consider more than just `input_ids` and `scores` in the future, and we should make sure that we don't need to rebuild them from the ground up when that happens. 

@patrickvonplaten what are your thoughts on changing e.g. [generation_utils.py#L451](https://github.com/huggingface/transformers/blob/5df79e2c41bf4b47ab4a36f903be163252714fe3/src/transformers/generation_utils.py#L451) when we need to also pass `cur_len` to enforce the max length or if we want the samples to look back at the `encoder_input_ids` ?

patrickvonplaten commented:
Thanks for the feedback! @yjernite - regarding your points: 

1) I think I see your point that `max_length` should be treated the same way as `min_length`. If we follow this appoarch, we would replace the `while cur_len < max_length:` with `while True:` and then break if `max_length` is hit. I'm a bit worried the people that will use one of the four functions directly, such as: 

```python 

pre_processor = # create your list of pre_processors here <= THIS MUST INCLUDE max_length
dist_warper = # create your dist warper here

outputs = model.sample(input_ids, pre_processor, dist_warper, max_length, pad_token_id, eos_token_id, **model_kwargs)
```

will forget to put a `max_length` "processor" in `pre_processor` and then the `while True:` loop would run forever. 

For me, the difference between `max_length` and and *e.g.* `min_length` is that `max_length` is a mandatory parameter for generation, which is why I left it as an input to the "sub" generation functions.
For you, what would be the big advantage of moving `max_length` to a preprocessor item (besides consistency?).

I'm 100% fine with extending the pre-processors or warpers (trying to not use the word samplers anymore :D) to accept more input arguments, but I think we could also do this in a future PR as it would not break backwards compatibility.

Happy to discuss what opinions the others have on this!

2) Agree - will have to see how to make that non-breaking, but should be possible!

3) Agree to add a `ModelOutputs` class to `generate()` that would include `attentions` and `hidden_states`. Regarding being able to backprop through `generate()` - I don't really think that this is super important. It would probably also require a lot of tweaking the way generate() is done currently so I'd prefer if ppl would just use the own fork/branch for these kind of things

4) This should already be possible I think. `decoder_input_ids` can be passed to generate.

patrickvonplaten committed:
make more tests pass

patrickvonplaten committed:
fix no bad words sampler

patrickvonplaten committed:
some fixes and tests for all distribution processors

patrickvonplaten committed:
fix test

patrickvonplaten committed:
fix rag slow tests

patrickvonplaten committed:
merge conflict

patrickvonplaten committed:
merge to master

patrickvonplaten committed:
add nograd to generate

patrickvonplaten committed:
make all slow tests pass

patrickvonplaten committed:
speed up generate

patrickvonplaten commented:
## Speed comparison:

sample search / greedy search yields equivalent results (TESTED on GPT2)

beam search yields ~5 % speed up thanks to the use of tensors instead of lists in `beam_scorer` (TESTED on BART and T5)

patrickvonplaten committed:
fix edge case bug

patrickvonplaten committed:
small fix

patrickvonplaten commented:
All slow tests that pass on master, also pass in this PR now. Ran all those tests: 

```
run_generation_integration_tests () {
        RUN_SLOW=1 pytest tests/test_modeling_pegasus.py
        RUN_SLOW=1 pytest tests/test_modeling_bart.py
        RUN_SLOW=1 pytest tests/test_modeling_t5.py
        RUN_SLOW=1 pytest tests/test_modeling_reformer.py
        RUN_SLOW=1 pytest tests/test_modeling_marian.py
        RUN_SLOW=1 pytest tests/test_modeling_mbart.py
        RUN_SLOW=1 pytest tests/test_modeling_prophetnet.py
        RUN_SLOW=1 pytest tests/test_modeling_xlm_prophetnet.py
        RUN_SLOW=1 pytest tests/test_modeling_encoder_decoder.py
        RUN_PIPELINE_TESTS=1 RUN_SLOW=1 pytest tests/test_pipelines_conversational.py
        RUN_PIPELINE_TESTS=1 RUN_SLOW=1 pytest tests/test_pipelines_text2text_generation.py
        RUN_PIPELINE_TESTS=1 RUN_SLOW=1 pytest tests/test_pipelines_text_generation.py
        RUN_PIPELINE_TESTS=1 RUN_SLOW=1 pytest tests/test_pipelines_summarization.py
        RUN_PIPELINE_TESTS=1 RUN_SLOW=1 pytest tests/test_pipelines_translation.py
        RUN_PIPELINE_TESTS=1 RUN_SLOW=1 pytest tests/test_pipelines_dialog.py
        RUN_SLOW=1 pytest tests/test_modeling_gpt2.py
        RUN_SLOW=1 pytest tests/test_modeling_xlnet.py
        RUN_SLOW=1 pytest tests/test_modeling_transfo_xl.py
        RUN_SLOW=1 pytest tests/test_modeling_rag.py
        RUN_SLOW=1 pytest tests/test_modeling_fsmt.py
        RUN_SLOW=1 pytest tests/test_modeling_blenderbot.py
}
```

Think I ran all of the important ones (cc @sshleifer )

patrickvonplaten committed:
correct typo

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -47,71 +57,194 @@ def adjust_logits_during_generation(self, logits, **kwargs):
+        is_pad_token_not_equal_to_eos_token_id = (eos_token_id is None) or (
+            (eos_token_id is not None) and (pad_token_id != eos_token_id)
+        )
+        if is_pad_token_in_inputs_ids and is_pad_token_not_equal_to_eos_token_id:
```
This new if statement should solve PR: #8032

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -47,71 +70,109 @@ def adjust_logits_during_generation(self, logits, **kwargs):
+        )
+        return decoder_input_ids
+
+    def _get_decoder_start_token_id(self, decoder_start_token_id, bos_token_id):
```
Yeah I see what you mean. I guess this would apply in 99% of the cases, but a problem could arrive if a user has defined `config.decoder_start_token_id` differently than `config.decoder.decoder_start_token_id` => this would then be overwritten at init.

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -47,71 +70,109 @@ def adjust_logits_during_generation(self, logits, **kwargs):
+            warpers.append(TemperatureSampler(temperature))
+        return warpers
+
+    def get_dist_pre_processor(
```
distribution? => I'm open for better namings!

thomwolf commented on src/transformers/generation_utils_dist_process.py:
```diff
@@ -0,0 +1,249 @@
+from torch.nn import functional as F
+
+
+class DistProcessorList(list):
```
I think I would also avoid the use of the abreviation `dist` which is often used for "distributed".

You can use `distributions` or `probability`

thomwolf reviewed (approved):
This looks really cool, amazing work @patrickvonplaten!

patrickvonplaten committed:
add type hints and docstrings

sgugger commented on src/transformers/generation_dist_process.py:
```diff
@@ -0,0 +1,312 @@
+        return scores
+
+
+class DistProcessor(ABC):
```
Nit: I'd put this class before the previous one, so I know what a `DistProcessor` is (not fan of the name) before encountering a `DistProcessorList`.

sgugger commented on src/transformers/generation_dist_process.py:
```diff
@@ -0,0 +1,312 @@
+
+class DistProcessorList(list):
+    """
+    This class can be used to create a list of :class:`~transformers.DistProcessor` to subsequently process a
```
:class:`~transformers.DistProcessor` requires that `DistProcessor` be put in the main `__init__` (which is not the case unless I missed something). If that's not the plan, you should use :class:`~transformers.generation_dist_process.DistProcessor`.

sgugger commented on src/transformers/generation_dist_process.py:
```diff
@@ -0,0 +1,312 @@
+
+
+class MinLengthDistProcessor(DistProcessor):
+    """DistProcessor enforcing a min-length by setting EOS probability to 0."""
```
All the `DistProcessor` subclasses should have a replace "DistProcessor" to :class:`~transformers.DistProcessor` (or :class:`~transformers.generation_dist_process.DistProcessor`) to make nice links in the doc :-)
At least if we plan to put those in the internal/generation.rst file. If not, you can ignore this comment (but I think we should).

It would also be super nice to document the arguments of each of those classes (even if we don't put them in the docs).

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -33,85 +43,234 @@ class GenerationMixin:
+        return model_kwargs
+
+    @staticmethod
+    def _reorder_cache(past: Tuple[torch.Tensor], beam_idx: torch.Tensor) -> Tuple[torch.Tensor]:
```
All  the previous private methods are very understandable from their names, this one a bit less. Could you add a comment/docstring?

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -33,85 +43,234 @@ class GenerationMixin:
+        self, top_k: int = None, top_p: float = None, temperature: float = None, num_beams: int = None
+    ) -> DistProcessorList:
+        """
+        This class returns a `DistProcessorList` object, that contains all distribution pre processing functions that
```
Use a proper reference to `DistProcessorList` and document arguments if possible.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -33,85 +43,234 @@ class GenerationMixin:
+        eos_token_id: int,
+    ) -> DistProcessorList:
+        """
+        This class returns a `DistProcessorList` object, that contains all distribution pre processing functions that
```
Same here.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -252,722 +405,472 @@ def generate(
+        pad_token_id: Optional[int] = None,
+        eos_token_id: Optional[int] = None,
+        **model_kwargs
+    ):
```
It would be great to have all public methods documented :-)

sgugger commented on src/transformers/generation_dist_process.py:
```diff
@@ -0,0 +1,312 @@
+        self.min_length = min_length
+        self.eos_token_id = eos_token_id
+
+    @add_start_docstrings_to_callable(DIST_PROCESSOR_INPUTS_DOCSTRING)
```
You don't need to put the docstring here again unless you add modifications, it will be inherited from the superclass.

sgugger commented on src/transformers/generation_dist_process.py:
```diff
@@ -0,0 +1,312 @@
+class DistProcessor(ABC):
+    """Abstract base class for all samplers which are probability distribution warps performed during generation."""
+
+    @add_start_docstrings_to_callable(DIST_PROCESSOR_INPUTS_DOCSTRING)
```
This method is for forward methods of PyTorch models, so you shouldn't use it in this case. `add_start_docstrings` is probably. We will rename it with @LysandreJik since you're not the first person to use it wrong ;-)

See also my comment below, if you end up never changing the docstrings, you could just define it once and for all here and not use any decorator.

sgugger reviewed (commented):
I don't know if this PR was ready for review or not but I still went ahead ;-). I still like this a lot, I added suggestions for the docs and I think it would be great if all your awesome work was documented. `GenerationMixin` is already documented in `main_classes/model` so all its public methods should have nice doc.

Then I think all you added warrants an `internal/generation` where we could document all the tools you added.

Last thing: I'm not super fan of `DistProcessor` as a name, mainly because I don't know that dist means distribution (I think distributed but I'm obsessed with Trainer ;-) ). `DistributionProcessor` might be a bit long. The class seems to be mainly doing some preprocessing on the logits though, so why not `LogitsProcessor`?

patrickvonplaten committed:
fix typos in tests

patrickvonplaten commented:
> I don't know if this PR was ready for review or not but I still went ahead ;-). I still like this a lot, I added suggestions for the docs and I think it would be great if all your awesome work was documented. `GenerationMixin` is already documented in `main_classes/model` so all its public methods should have nice doc.
> 
> Then I think all you added warrants an `internal/generation` where we could document all the tools you added.
> 
> Last thing: I'm not super fan of `DistProcessor` as a name, mainly because I don't know that dist means distribution (I think distributed but I'm obsessed with Trainer ;-) ). `DistributionProcessor` might be a bit long. The class seems to be mainly doing some preprocessing on the logits though, so why not `LogitsProcessor`?

Changed to `LogitsProcessor` - like that name! Yeah the docs weren't ready yet, but thanks for the feedback :-)

patrickvonplaten committed:
add beam search tests

patrickvonplaten committed:
add tests for beam scorer

patrickvonplaten committed:
fix test rag

yjernite commented:
Hmmm OK so after thinking about it for a bit I realize that having a `max_length` logits processor would change the current beam search behavior in some cases.

Currently, the beam gets re-ordered at `max_length` without penalizing sequences that aren't "well-formed" (haven't reached EOS within the allotted time). If we switched to a `LogitsProcessor` however, the beam search would up-weight sequences that were more likely to lead to an EOS at time step (`max_length`-1)

I think this new behavior is better and definitely would like the option to implement it (@srush would love your input on that especially), but I'm not sure how breaking it is (conversely, it might also account for some of the different scores we've had from other libs). We can also just add a `BeamSearchMaxLengthLogitsProcessor` now or later.

What do you think @thomwolf @patrickvonplaten @sshleifer ?

sshleifer commented:
> We can also just add a BeamSearchMaxLengthLogitsProcessor now or later.

This sounds useful! Bart forces all the generations of length `max_length -1` to end in EOS, which helps performance. Might help other models too.  Would lean towards adding it later.

patrickvonplaten committed:
finish beam search tests

patrickvonplaten committed:
move generation tests in seperate file

patrickvonplaten committed:
fix generation tests

patrickvonplaten committed:
more tests

patrickvonplaten committed:
add aggressive generation tests

patrickvonplaten committed:
fix tests

patrickvonplaten committed:
add gpt2 sample test

patrickvonplaten committed:
add more docstring

patrickvonplaten committed:
Merge branch 'master' into generation_refactor

patrickvonplaten committed:
add more docs

patrickvonplaten committed:
finish doc strings

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +334,812 @@ def generate(
-        else:
-            batch_size = 1
-
-        assert isinstance(max_length, int) and max_length > 0, "`max_length` should be a strictly positive integer."
```
all those assert statements are moved to the corresponding `LogitProcessor` / `LogitWarper`

patrickvonplaten committed:
apply some more of sylvains and sams comments

patrickvonplaten committed:
fix some typos

patrickvonplaten committed:
make fix copies

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+        kwargs:
+            Additional kwargs including special tokens, such as :obj:`eos_token_id` and :obj:`pad_token_id`.
+
+    Return: :obj:`UserDict` composed of the fields as defined above:
```
As you can see [here](https://110256-155220641-gh.circle-artifacts.com/0/docs/_build/html/internal/generation_utils.html#beamsearch), this is not rendered properly in the docs.
```suggestion
    Return:
        :obj:`UserDict`: A dictionary composed of the fields as defined above:
```
The last colon might make sphinx bug in which case we should remove it.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+        next_tokens (:obj:`torch.LongTensor` of shape :obj:`(batch_size, 2 * num_beams)`):
+            :obj:`input_ids` of the tokens corresponding to the top :obj:`2 * num_beams` non-finished beam hypotheses.
+        next_indices (:obj:`torch.LongTensor` of shape :obj:`(batch_size, 2 * num_beams)`):
+            beam indices indicating to which beam hypothesis the :obj:`next_tokens` correspond.
```
```suggestion
            Beam indices indicating to which beam hypothesis the :obj:`next_tokens` correspond.
```
Mega-nit

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+        next_indices (:obj:`torch.LongTensor` of shape :obj:`(batch_size, 2 * num_beams)`):
+            beam indices indicating to which beam hypothesis the :obj:`next_tokens` correspond.
+        kwargs:
+            Additional kwargs including special tokens, such as :obj:`eos_token_id` and :obj:`pad_token_id`.
```
```suggestion
            Additional keyword arguments including special tokens, such as :obj:`eos_token_id` and :obj:`pad_token_id`.
```
Additional keyword arguments passed along to who? It would be great to expand on that.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+
+    Return: :obj:`UserDict` composed of the fields as defined above:
+
+            - next_beam_scores (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- updated scores of
```
```suggestion
            - **next_beam_scores** (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- Updated scores of
```

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+
+            - next_beam_scores (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- updated scores of
+              all non-finished beams.
+            - next_beam_tokens (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- next tokens to be
```
```suggestion
            - **next_beam_tokens** (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- Next tokens to be
```

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+              all non-finished beams.
+            - next_beam_tokens (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- next tokens to be
+              added to the non-finished beam_hypotheses.
+            - next_beam_indices (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- beam indices
```
```suggestion
            - **next_beam_indices** (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`) -- Beam indices
```

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+        final_beam_indices (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_beams)`):
+            The beam indices indicating to which beam the :obj:`final_beam_tokens` shall be added.
+        kwargs:
+            Additional kwargs including special tokens, such as :obj:`eos_token_id` and :obj:`pad_token_id`.
```
```suggestion
            Additional keyword arguments including special tokens, such as :obj:`eos_token_id` and :obj:`pad_token_id`.
```
Same as above. Where do they go?

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+        num_beams (:obj:`int`):
+            Number of beams for beam search.
+        device (:obj:`torch.device`):
+            Defines the device type (*e.g.*, :obj:`"cpu"` or :obj:`"cuda"`) on which `BeamSearchScorer` will be
```
```suggestion
            Defines the device type (*e.g.*, :obj:`"cpu"` or :obj:`"cuda"`) on which this instance of :obj:`BeamSearchScorer` will be
```

sgugger commented on src/transformers/generation_logits_process.py:
```diff
@@ -0,0 +1,374 @@
+
+class NoRepeatNGramLogitsProcessor(LogitsProcessor):
+    r"""
+    :class:`transformers.LogitsProcessor` that enforces no repetition of n-grams. See Fairseq:
```
Can't suggest on two lines but use a link here: `Fairseq <...>`__

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
-            outputs = model.generate(input_ids=input_ids, num_beams=5, num_return_sequences=3, temperature=1.5)  # generate 3 independent sequences using beam search decoding (5 beams) with sampling from initial context 'The dog'
-            for i in range(3): #  3 output sequences were generated
-                print('Generated {}: {}'.format(i, tokenizer.decode(outputs[i], skip_special_tokens=True)))
+            tokenizer = AutoTokenizer.from_pretrained('t5-base')   # Initialize tokenizer
```
Can we put the comments on the lines before so that it all gets below 119 chars?

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+            eos_token_id (:obj:`int`, `optional`):
+                The id of the `end-of-sequence` token.
+            model_kwargs:
+                Additional model specific kwargs will be forwarded to the :obj:`forward` function of the model. If
```
```suggestion
                Additional model specific keyword arguments will be forwarded to the :obj:`forward` function of the model. If
```

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+
+            input_prompt = "Today is a beautiful day, and"
+            input_ids = tokenizer(input_prompt, return_tensors="pt").input_ids
+            logits_processor = LogitsProcessorList([MinLengthLogitsProcessor(15, eos_token_id=model.config.eos_token_id)])  # instantiate logits processors
```
Same here, please put the comment above.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+
+        Examples::
+
+            from transformers import AutoTokenizer, AutoModelForCausalLM, LogitsProcessorList, MinLengthLogitsProcessor, TopKLogitsWarper, TemperatureLogitsWarper
```
Can we format the import black-style with parenthesis (and avoid going over 119 chars :-) )

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+            input_prompt = "Today is a beautiful day, and"
+            input_ids = tokenizer(input_prompt, return_tensors="pt").input_ids
+
+            logits_processor = LogitsProcessorList([MinLengthLogitsProcessor(15, eos_token_id=model.config.eos_token_id)])  # instantiate logits processors
```
As before, please put the comments above.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+
+        Examples::
+
+            from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, LogitsProcessorList, MinLengthLogitsProcessor, BeamSearchScorer
```
Same here, those imports could be done on several lines.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+                num_beams=num_beams,
+                device=model.device,
+            )
+            logits_processor = LogitsProcessorList([MinLengthLogitsProcessor(5, eos_token_id=model.config.eos_token_id)])  # instantiate logits processors
```
Another line too long ;-)

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
-    @staticmethod
-    def _reorder_cache(past: Tuple, beam_idx: Tensor) -> Tuple[Tensor]:
-        return tuple(layer_past.index_select(1, beam_idx) for layer_past in past)
+            from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, LogitsProcessorList, MinLengthLogitsProcessor, TopKLogitsWarper, TemperatureLogitsWarper, BeamSearchScorer
```
Those imports should be done on several lines too.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+                num_beams=num_beams,
+                device=model.device,
             )
+            logits_processor = LogitsProcessorList([MinLengthLogitsProcessor(5, eos_token_id=model.config.eos_token_id)])  # instantiate logits processors
```
Same, please put the comment above.

sgugger reviewed (approved):
This all looks great to me. I have final nits on the docs, some general rules not to forget:
- no abbreviation in the documentation as we have all the space we want to explain things to the user
- no lines > 119 pretty please, the script takes care of everything except the examples and some of the examples have veeeeeery long lines.

LysandreJik commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+# coding=utf-8
+# Copyright 2020 HuggingFace Inc.
```
```suggestion
# Copyright 2020 The HuggingFace Inc. team
```

LysandreJik commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+        input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size * num_beams, sequence_length)`):
+            Indices of input sequence tokens in the vocabulary.
+
+            Indices can be obtained using :class:`~transformers.BertTokenizer`. See
```
```suggestion
            Indices can be obtained using any class inheriting from :class:`~transformers.PreTrainedTokenizer`. See
```

LysandreJik commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+        input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size * num_beams, sequence_length)`):
+            Indices of input sequence tokens in the vocabulary.
+
+            Indices can be obtained using :class:`~transformers.BertTokenizer`. See
```
```suggestion
            Indices can be obtained using any class inheriting from :class:`~transformers.PreTrainedTokenizer`. See
```

LysandreJik commented on src/transformers/generation_beam_search.py:
```diff
@@ -0,0 +1,356 @@
+            Exponential penalty to the length. 1.0 means no penalty. Set to values < 1.0 in order to encourage the
+            model to generate shorter sequences, to a value > 1.0 in order to encourage the model to produce longer
+            sequences.
+        do_early_stopping (:obj:`bool`, `optional`, defaults to :obj:`False`):
```
```suggestion
        do_early_stopping (:obj:`bool`, `optional`, defaults to :obj:`True`):
```

patrickvonplaten replied:
Nice catch!

LysandreJik commented on src/transformers/generation_logits_process.py:
```diff
@@ -0,0 +1,374 @@
+# coding=utf-8
+# Copyright 2020 HuggingFace Inc.
```
```suggestion
# Copyright 2020 The HuggingFace Inc. team
```

LysandreJik commented on src/transformers/generation_logits_process.py:
```diff
@@ -0,0 +1,374 @@
+
+    def _set_scores_to_inf_for_banned_tokens(self, scores: torch.Tensor, banned_tokens: List[List[int]]) -> None:
+        """
+        Modifies the scores in place by setting the banned token positions to `-inf`. Banned token is expected to be a
+        list of list of banned tokens to ban in the format [[batch index, vocabulary position],...
+
+            Args:
+                scores: logits distribution of shape (batch size, vocabulary size)
+                banned_tokens: list of list of tokens to ban of length (batch_size)
```
I think there's an indentation issue here

LysandreJik commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
         Examples::
 
             tokenizer = AutoTokenizer.from_pretrained('distilgpt2')   # Initialize tokenizer
-            model = AutoModelWithLMHead.from_pretrained('distilgpt2')    # Download model and configuration from S3 and cache.
+            model = AutoModelForCausalLM.from_pretrained('distilgpt2')    # Download model and configuration from S3 and cache.
```
(nit) Since you're editing these examples, mind adding the import statement as well? Will be one less code sample to debug when we turn the doctests back on :)

LysandreJik commented on src/transformers/generation_utils.py:
```diff
@@ -182,792 +343,759 @@ def generate(
+                The sequence used as a prompt for the generation. If :obj:`None` the method initializes it as an empty
+                :obj:`torch.LongTensor` of shape :obj:`(1,)`.
+            beam_scorer (:obj:`BeamScorer`):
+                An derived instance of :class:`~transformers.BeamScorer` that defines how beam hypotheses are
```
```suggestion
                An derived instance of :class:`~transformers.BeamScorer` that defines how beam hypotheses are
```
```suggestion
                A derived instance of :class:`~transformers.BeamScorer` that defines how beam hypotheses are
```

LysandreJik commented on src/transformers/modeling_rag.py:
```diff
@@ -1388,64 +1389,57 @@ def extend_enc_output(tensor, num_beams=None):
+                **model_kwargs,
             )
+        else:
+            raise ValueError(f"`num_beams` has to be strictly positive (> 0), but is {num_beams}")
```
(nittiest of nits)
```suggestion
            raise ValueError(f"`num_beams` has to be an integer strictly superior to 0 (≥ 1), but is {num_beams}")
```

LysandreJik reviewed (approved):
Tremendous job on this PR! The code looks great, way easier to read and understand.
Thanks for taking the time to do this!

Only left some nits related to documentation.

patrickvonplaten commented:
> This all looks great to me. I have final nits on the docs, some general rules not to forget:
> 
> * no abbreviation in the documentation as we have all the space we want to explain things to the user
> * no lines > 119 pretty please, the script takes care of everything except the examples and some of the examples have veeeeeery long lines.



> This all looks great to me. I have final nits on the docs, some general rules not to forget:
> 
> * no abbreviation in the documentation as we have all the space we want to explain things to the user
> * no lines > 119 pretty please, the script takes care of everything except the examples and some of the examples have veeeeeery long lines.

Noo, broke the 119 rule again - I though the script would now save me from everything :D Will correct this! Should I also add the `>>>` for the example code or is this just for the model's forward function?

sgugger commented:
> Should I also add the `>>>` for the example code or is this just for the model's forward function?

It's only used as a marker for doctests, so you should do this for examples that are not slow. We still need to resuscitate the doctests with @LysandreJik for it to have any use, though.

LysandreJik commented:
The doctest run as slow tests so you can still add them even if they're slow!

patrickvonplaten committed:
apply lysandres and sylvains comments

patrickvonplaten committed:
final corrections on examples

patrickvonplaten commented:
> The doctest run as slow tests so you can still add them even if they're slow!

Okey added `>>>` to all examples and made them pretty

patrickvonplaten committed:
small fix for reformer

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -189,6 +189,8 @@ def _update_model_kwargs_for_generation(
         elif "mems" in outputs:
             model_kwargs["past"] = outputs.mems
+        elif "past_buckets_states" in outputs:
+            model_kwargs["past"] = outputs.past_buckets_states
```
credit goes to @guillaume-be for catching this bug in https://github.com/huggingface/transformers/pull/8252

luohongyin commented:
Hi, thanks for creating this PR and the code looks great! Is it possible to decode and meanwhile return the token probabilities? It would be very helpful in the following scenarios -

I. Calculating perplexities of generated texts
II. Reinforcement learning for text generation.

Sai-Ashish commented:
Is it possible to backpropagate the conditional text generation model parameters on the loss defined via this .generate() function?

rafikg commented:
@patrickvonplaten can you share how to calculate the output probabilities of each token given by generate() ?
